### PR TITLE
Backend: Create a caching mechanism to cache API requests

### PR DESCRIPTION
**Pull request by JakWai01**
Closes https://github.com/alphahorizonio/consumat.io/issues/13

### CODE REVIEW

File: consumatio/external/tmdb.py [hunk 1]
```diff
@@ -9,47 +9,68 @@
...
\ No newline at end of file
+        query = f'https://api.themoviedb.org/3/tv/{tv_id}/season/{season_number}/episode/{episode_number}/images?api_key=c4bd02adee6e73c4d17e69b039267c90'
+        data = get_data(query)
+        return episode_images_to_dict(data)
+
+def get_data(query):
+   db = Database()
+
+   if (db.is_cached(query)):
+      data = eval(db.get_from_cache(query))
```
Comment (r2binx) [inline consumatio/external/tmdb.py:72]: `eval()` can be quite dangerous. Instead please consider using `json.loads()`
```suggestion
      data = json.loads(db.get_from_cache(query))
```

File: consumatio/external/tmdb.py [hunk 2]
```diff
@@ -9,47 +9,68 @@
...
+   else:
+      data = requests.get(query).json() 
+      db.cache(query, str(data))
```
Comment (r2binx) [inline consumatio/external/tmdb.py:75]: of course we would need then to change `str()` to `json.dumps()` if we're going the json route. 
```suggestion
      db.cache(query, json.dumps(data))
```

File: consumatio/external/db.py [hunk 3]
```diff
@@ -0,0 +1,47 @@
+import sqlite3
+import datetime
+
+class Database():
+
+    def cache(self, query, body):
+        con = sqlite3.connect('db.sqlite3')
+        cur = con.cursor()
+
+        cur.execute('''CREATE TABLE IF NOT EXISTS cache(query text UNIQUE, body text, last_changed date)''')
```
Comment (r2binx) [inline consumatio/external/db.py:10]: it would make sense to move this into a constructor.

File: consumatio/external/tmdb.py [hunk 4]
```diff
@@ -9,47 +9,68 @@
...
     def get_episode_images(self, tv_id, season_number, episode_number):
-        data = requests.get(f'https://api.themoviedb.org/3/tv/{tv_id}/season/{season_number}/episode/{episode_number}/images?api_key=c4bd02adee6e73c4d17e69b039267c90').json()
-        return episode_images_to_dict(data)
\ No newline at end of file
+        query = f'https://api.themoviedb.org/3/tv/{tv_id}/season/{season_number}/episode/{episode_number}/images?api_key=c4bd02adee6e73c4d17e69b039267c90'
+        data = get_data(query)
+        return episode_images_to_dict(data)
+
+def get_data(query):
+   db = Database()
```
Comment (r2binx) [inline consumatio/external/tmdb.py:69]: IMO creating a database instance for every query isn't necessary in our case.

**Review by r2binx — CHANGES_REQUESTED** [review]
Thanks for your contribution! Except for my few nitpicks it looks good! We might need to discuss if sqlite is the right solution in this case but for now it's alright.

**Comment by r2binx** [comment]
Regarding SQLite my main concern is the issue of [locking](https://www.sqlite.org/lockingv3.html) when writing to the "cache". 
I was thinking of a more primitive way of caching, maybe something like [Flask-Cache](https://pythonhosted.org/Flask-Cache/) or a similar file system based approach would be more appropriate.
Maybe @pojntfx can chime in here ✌️

**Review by r2binx — APPROVED** [review]

File: consumatio/external/tmdb.py [hunk 5]
```diff
@@ -9,47 +9,70 @@
...
 
 class Tmdb():
-    
+
+    def __init__(self):
+        self.db = Database()
+
     def get_movie_details(self, movie_id):
-        data = requests.get(f'https://api.themoviedb.org/3/movie/{movie_id}?api_key=c4bd02adee6e73c4d17e69b039267c90&language=en-US').json()
+        query = f'https://api.themoviedb.org/3/movie/{movie_id}?api_key=c4bd02adee6e73c4d17e69b039267c90&language=en-US'
```
Comment (pojntfx) [inline consumatio/external/tmdb.py:21]: Please decompose `https://api.themoviedb.org/3/` here and in the following API requests and get the `api_key` from an env variable

**Review by pojntfx — CHANGES_REQUESTED** [review]
NIce; please take a look at the comment, but this should be a good PoC IMHO.

**Comment by pojntfx** [comment]
@r2binx I agree with your point - SQLite is really fast when it comes to read operations, but due to the global write lock it is probably quite slow here as the cache is probably quite write-heavy, especially when it comes to search results etc. Maybe a) create a write memcache in the future or b) use a caching mechanism that doesn't require locking, i.e. something like an embedded version of redis.

File: consumatio/external/tmdb.py (hunk 5 above)
Comment (pojntfx) [reply consumatio/external/tmdb.py:21]: I see that there is already an issue for the API keys, closing

**Review by pojntfx — APPROVED** [review]

**Pull request merged**